aws/request/request_retry_test.go

```diff
@@ -0,0 +1,107 @@
+package request
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+	"time"
+)
+
+func r(t *testing.T, url string) *http.Request {
+	r, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		t.Fatalf("can't forge request: %v", err)
+	}
+	return r
+}
+
+func Test_shouldRetryCancel_nil(t *testing.T) {
+	shouldRetryCancel(nil)
+}
+
+func Test_shouldRetryCancel_timeout(t *testing.T) {
+
+	tr := &http.Transport{}
+	defer tr.CloseIdleConnections()
+	cli := http.Client{
+		Timeout:   time.Nanosecond,
+		Transport: tr,
+	}
+
+	resp, err := cli.Do(r(t, "https://179.179.179.179/no/such/host"))
+	if resp != nil {
+		resp.Body.Close()
+	}
+	if err == nil {
+		t.Fatal("This should have failed.")
+	}
+	debugerr(t, err)
+
+	if shouldRetryCancel(err) == false {
+		t.Errorf("this request timed out and should be retried")
+	}
+}
+
+func Test_shouldRetryCancel_cancelled(t *testing.T) {
+
+	tr := &http.Transport{}
+	defer tr.CloseIdleConnections()
+	cli := http.Client{
+		Transport: tr,
+	}
+
+	unblockc := make(chan bool)
+	srvr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		fmt.Fprintf(w, "Hello")
+		w.(http.Flusher).Flush() // send headers and some body
+		<-unblockc               // block forever
+	}))
+	defer srvr.Close()
+	defer close(unblockc)
+
+	r := r(t, srvr.URL)
+	ch := make(chan struct{})
+	r.Cancel = ch
+	close(ch) // request is cancelled before anything
+
+	resp, err := cli.Do(r)
+	if resp != nil {
+		resp.Body.Close()
+	}
+	if err == nil {
+		t.Fatal("This should have failed.")
+	}
+
+	debugerr(t, err)
+
+	if shouldRetryCancel(err) == true {
+		t.Errorf("this request was cancelled and should not be retried")
+	}
+}
+
+func debugerr(t *testing.T, err error) {
+	type temporaryError interface {
```

Comment:
I don't think this local definition of the interface is needed since the unit test is within the `request` package. The `temporary` type should be accessible.

Reply:
oh, right, I added this at different moments ! Good catch.